libdnf/base/transaction.cpp

```diff
@@ -466,9 +416,44 @@ Transaction::TransactionRunResult Transaction::Impl::run(
     if (ret == 0) {
         // set the new system state
         auto & system_state = base->get_system_state();
-        for (const auto & tspkg : packages) {
-            system_state.set_reason(tspkg.get_package().get_na(), tspkg.get_reason());
+
+        rpm::PackageQuery installed_query(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES);
+        installed_query.filter_installed();
+
+        // Iterate in reverse, inbound actions are first in the vector, we want to process outbound first
+        for (auto it = packages.rbegin(); it != packages.rend(); ++it) {
+            auto tspkg = *it;
+            if (transaction_item_action_is_inbound(tspkg.get_action())) {
+                system_state.set_package_reason(tspkg.get_package().get_na(), tspkg.get_reason());
+                system_state.set_package_from_repo(tspkg.get_package().get_nevra(), tspkg.get_package().get_repo_id());
+            } else if (transaction_item_action_is_outbound(tspkg.get_action())) {
+                // Check if the NA is still installed. We do this check for all outbound actions.
+                //
+                // For REMOVE, if there's a package with the same NA still on the system,
+                // it's an installonly package and we're keeping the reason.
+                //
+                // For REPLACED, we don't know if it was UPGRADE/DOWNGRADE/REINSTALL
+                // (we're keeping the reason) or it's an obsolete (we're removing the reason)
+
+                // We need to filter out packages that are being removed in the transaction
+                // (the installed query still contains the packages before this transaction)
+                installed_query.filter_nevra({tspkg.get_package().get_nevra()}, libdnf::sack::QueryCmp::NEQ);
+
+                rpm::PackageQuery query(installed_query);
+                query.filter_name({tspkg.get_package().get_name()});
+                query.filter_arch({tspkg.get_package().get_arch()});
```

Comment:
I am guessing now. In case I have 2 installonly package and I am removing both of them I guess that the code will keep the reason.

Reply:
Anyway your approach is in principle better then the previous logic. But we have to resolve all corner cases.

Reply:
It works for this case, I've now enabled a scenario that tests it.

Reply:
Perfect, I realized one thing byt it is a performance optimization. We have a filter that can replace 
```
                query.filter_name({tspkg.get_package().get_name()});
                query.filter_arch({tspkg.get_package().get_arch()});
```
See 
```

    /// Filter packages by their `name` and `arch` based on names and arches of the packages in the `package_set`.
    ///
    /// @param package_set      PackageSet with Package objects the filter is matched against.
    /// @param cmp              A comparison (match) operator, defaults to `QueryCmp::EQ`.
    ///                         Supported values: `EQ`, `NEQ`.
    /// @since 5.0
    void filter_name_arch(const PackageSet & package_set, libdnf::sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
```

If you have only package as an argument we can create a variant for the package.

Reply:
So far there's no filter for a single package on `PackageQuery` interface. I wouldn't add one randomly, if we want to have filters on `Package`s, I'd add them systematically.